{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Set up notebook environment\n",
    "## NOTE: Use a QIIME2 kernel"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 25,
   "metadata": {},
   "outputs": [],
   "source": [
    "import os\n",
    "import biom\n",
    "import warnings\n",
    "import pickle\n",
    "import numpy as np\n",
    "import pandas as pd\n",
    "import qiime2 as q2\n",
    "from biom import Table\n",
    "from skbio import OrdinationResults\n",
    "from skbio.stats import subsample_counts\n",
    "from skbio.stats.distance import permanova, anosim, mantel\n",
    "from skbio.stats.distance import DistanceMatrix\n",
    "from qiime2.plugins.deicode.actions import rpca\n",
    "from qiime2.plugins.feature_table.actions import rarefy\n",
    "from qiime2.plugins.diversity.actions import beta_group_significance\n",
    "from qiime2.plugins.emperor.actions import biplot, plot\n",
    "from qiime2.plugins.diversity.actions import (beta,\n",
    "                                              beta_phylogenetic,\n",
    "                                              pcoa)\n",
    "from qiime2.plugins import demux, deblur, quality_filter, \\\n",
    "                           metadata, feature_table, alignment, \\\n",
    "                           phylogeny, diversity, emperor, feature_classifier, \\\n",
    "                           taxa, composition\n",
    "\n",
    "from assets.step_wise_anova import run_stepwise_anova\n",
    "from qiime2.plugins.fragment_insertion.actions import filter_features\n",
    "warnings.filterwarnings(\"ignore\", category=DeprecationWarning)\n",
    "\n",
    "# helper functions\n",
    "from assets.util import (mantel_matched, simulate_depth,\n",
    "                        all_dists, nested_permanova)\n",
    "\n",
    "# plotting\n",
    "import matplotlib\n",
    "import seaborn as sns\n",
    "import matplotlib.pyplot as plt\n",
    "warnings.filterwarnings('ignore')\n",
    "\n",
    "plt.style.use('ggplot')\n",
    "%matplotlib inline\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Subset metadata to make paired files between extraction kits"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 26,
   "metadata": {},
   "outputs": [],
   "source": [
    "SEQ_TYPE='shotgun'"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 27,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>sample_name</th>\n",
       "      <th>bacteria</th>\n",
       "      <th>bacteria_function</th>\n",
       "      <th>biomass</th>\n",
       "      <th>buffer_removal</th>\n",
       "      <th>col</th>\n",
       "      <th>collection_timestamp</th>\n",
       "      <th>compressed_well</th>\n",
       "      <th>description</th>\n",
       "      <th>diet</th>\n",
       "      <th>...</th>\n",
       "      <th>sample_type2</th>\n",
       "      <th>scientific_name</th>\n",
       "      <th>sex</th>\n",
       "      <th>storage_solution</th>\n",
       "      <th>swab_type</th>\n",
       "      <th>taxon_id</th>\n",
       "      <th>title</th>\n",
       "      <th>vivarium</th>\n",
       "      <th>water</th>\n",
       "      <th>replicate_id</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>14332.361158597</td>\n",
       "      <td>not applicable</td>\n",
       "      <td>not applicable</td>\n",
       "      <td>low</td>\n",
       "      <td>dry</td>\n",
       "      <td>8.0</td>\n",
       "      <td>3/24/22</td>\n",
       "      <td>O8</td>\n",
       "      <td>skin_hand_right</td>\n",
       "      <td>not applicable</td>\n",
       "      <td>...</td>\n",
       "      <td>skin_hand_right</td>\n",
       "      <td>human skin metagenome</td>\n",
       "      <td>female</td>\n",
       "      <td>isopropanol</td>\n",
       "      <td>swube</td>\n",
       "      <td>539655.0</td>\n",
       "      <td>Matrix_pipeline_validation</td>\n",
       "      <td>not applicable</td>\n",
       "      <td>not applicable</td>\n",
       "      <td>D skin_hand_right</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>14332.361162759</td>\n",
       "      <td>not applicable</td>\n",
       "      <td>not applicable</td>\n",
       "      <td>low</td>\n",
       "      <td>dry</td>\n",
       "      <td>3.0</td>\n",
       "      <td>3/24/22</td>\n",
       "      <td>E3</td>\n",
       "      <td>skin_armpit_right</td>\n",
       "      <td>not applicable</td>\n",
       "      <td>...</td>\n",
       "      <td>skin_armpit_right</td>\n",
       "      <td>human skin metagenome</td>\n",
       "      <td>male</td>\n",
       "      <td>etoh</td>\n",
       "      <td>swube</td>\n",
       "      <td>539655.0</td>\n",
       "      <td>Matrix_pipeline_validation</td>\n",
       "      <td>not applicable</td>\n",
       "      <td>not applicable</td>\n",
       "      <td>B skin_armpit_right</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>14332.361162761</td>\n",
       "      <td>not applicable</td>\n",
       "      <td>not applicable</td>\n",
       "      <td>low</td>\n",
       "      <td>dry</td>\n",
       "      <td>5.0</td>\n",
       "      <td>3/24/22</td>\n",
       "      <td>G5</td>\n",
       "      <td>skin_armpit_right</td>\n",
       "      <td>not applicable</td>\n",
       "      <td>...</td>\n",
       "      <td>skin_armpit_right</td>\n",
       "      <td>human skin metagenome</td>\n",
       "      <td>female</td>\n",
       "      <td>etoh</td>\n",
       "      <td>swube</td>\n",
       "      <td>539655.0</td>\n",
       "      <td>Matrix_pipeline_validation</td>\n",
       "      <td>not applicable</td>\n",
       "      <td>not applicable</td>\n",
       "      <td>C skin_armpit_right</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>11</th>\n",
       "      <td>14332.361162773</td>\n",
       "      <td>not applicable</td>\n",
       "      <td>not applicable</td>\n",
       "      <td>low</td>\n",
       "      <td>dry</td>\n",
       "      <td>4.0</td>\n",
       "      <td>3/24/22</td>\n",
       "      <td>C4</td>\n",
       "      <td>skin_hand_right</td>\n",
       "      <td>not applicable</td>\n",
       "      <td>...</td>\n",
       "      <td>skin_hand_right</td>\n",
       "      <td>human skin metagenome</td>\n",
       "      <td>male</td>\n",
       "      <td>etoh</td>\n",
       "      <td>swube</td>\n",
       "      <td>539655.0</td>\n",
       "      <td>Matrix_pipeline_validation</td>\n",
       "      <td>not applicable</td>\n",
       "      <td>not applicable</td>\n",
       "      <td>B skin_hand_right</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>21</th>\n",
       "      <td>14332.361162788</td>\n",
       "      <td>not applicable</td>\n",
       "      <td>not applicable</td>\n",
       "      <td>low</td>\n",
       "      <td>dry</td>\n",
       "      <td>6.0</td>\n",
       "      <td>3/24/22</td>\n",
       "      <td>G6</td>\n",
       "      <td>skin_hand_right</td>\n",
       "      <td>not applicable</td>\n",
       "      <td>...</td>\n",
       "      <td>skin_hand_right</td>\n",
       "      <td>human skin metagenome</td>\n",
       "      <td>female</td>\n",
       "      <td>etoh</td>\n",
       "      <td>swube</td>\n",
       "      <td>539655.0</td>\n",
       "      <td>Matrix_pipeline_validation</td>\n",
       "      <td>not applicable</td>\n",
       "      <td>not applicable</td>\n",
       "      <td>C skin_hand_right</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>235</th>\n",
       "      <td>14332.363197876</td>\n",
       "      <td>not applicable</td>\n",
       "      <td>not applicable</td>\n",
       "      <td>low</td>\n",
       "      <td>dry</td>\n",
       "      <td>8.0</td>\n",
       "      <td>3/24/22</td>\n",
       "      <td>H8</td>\n",
       "      <td>skin_hand_left</td>\n",
       "      <td>not applicable</td>\n",
       "      <td>...</td>\n",
       "      <td>skin_hand_left</td>\n",
       "      <td>human skin metagenome</td>\n",
       "      <td>female</td>\n",
       "      <td>isopropanol</td>\n",
       "      <td>custom</td>\n",
       "      <td>539655.0</td>\n",
       "      <td>Matrix_pipeline_validation</td>\n",
       "      <td>not applicable</td>\n",
       "      <td>not applicable</td>\n",
       "      <td>D skin_hand_left</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>236</th>\n",
       "      <td>14332.363197877</td>\n",
       "      <td>not applicable</td>\n",
       "      <td>not applicable</td>\n",
       "      <td>low</td>\n",
       "      <td>dry</td>\n",
       "      <td>2.0</td>\n",
       "      <td>3/23/22</td>\n",
       "      <td>J2</td>\n",
       "      <td>skin_hand_left</td>\n",
       "      <td>not applicable</td>\n",
       "      <td>...</td>\n",
       "      <td>skin_hand_left</td>\n",
       "      <td>human skin metagenome</td>\n",
       "      <td>female</td>\n",
       "      <td>isopropanol</td>\n",
       "      <td>custom</td>\n",
       "      <td>539655.0</td>\n",
       "      <td>Matrix_pipeline_validation</td>\n",
       "      <td>not applicable</td>\n",
       "      <td>not applicable</td>\n",
       "      <td>A skin_hand_left</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>240</th>\n",
       "      <td>14332.363197882</td>\n",
       "      <td>not applicable</td>\n",
       "      <td>not applicable</td>\n",
       "      <td>low</td>\n",
       "      <td>dry</td>\n",
       "      <td>7.0</td>\n",
       "      <td>3/24/22</td>\n",
       "      <td>D7</td>\n",
       "      <td>skin_armpit_left</td>\n",
       "      <td>not applicable</td>\n",
       "      <td>...</td>\n",
       "      <td>skin_armpit_left</td>\n",
       "      <td>human skin metagenome</td>\n",
       "      <td>female</td>\n",
       "      <td>etoh</td>\n",
       "      <td>custom</td>\n",
       "      <td>539655.0</td>\n",
       "      <td>Matrix_pipeline_validation</td>\n",
       "      <td>not applicable</td>\n",
       "      <td>not applicable</td>\n",
       "      <td>D skin_armpit_left</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>243</th>\n",
       "      <td>14332.363197890</td>\n",
       "      <td>not applicable</td>\n",
       "      <td>not applicable</td>\n",
       "      <td>low</td>\n",
       "      <td>dry</td>\n",
       "      <td>8.0</td>\n",
       "      <td>3/24/22</td>\n",
       "      <td>D8</td>\n",
       "      <td>skin_hand_left</td>\n",
       "      <td>not applicable</td>\n",
       "      <td>...</td>\n",
       "      <td>skin_hand_left</td>\n",
       "      <td>human skin metagenome</td>\n",
       "      <td>female</td>\n",
       "      <td>etoh</td>\n",
       "      <td>custom</td>\n",
       "      <td>539655.0</td>\n",
       "      <td>Matrix_pipeline_validation</td>\n",
       "      <td>not applicable</td>\n",
       "      <td>not applicable</td>\n",
       "      <td>D skin_hand_left</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>245</th>\n",
       "      <td>14332.363197892</td>\n",
       "      <td>not applicable</td>\n",
       "      <td>not applicable</td>\n",
       "      <td>low</td>\n",
       "      <td>dry</td>\n",
       "      <td>7.0</td>\n",
       "      <td>3/24/22</td>\n",
       "      <td>F7</td>\n",
       "      <td>skin_armpit_left</td>\n",
       "      <td>not applicable</td>\n",
       "      <td>...</td>\n",
       "      <td>skin_armpit_left</td>\n",
       "      <td>human skin metagenome</td>\n",
       "      <td>female</td>\n",
       "      <td>etoh</td>\n",
       "      <td>custom</td>\n",
       "      <td>539655.0</td>\n",
       "      <td>Matrix_pipeline_validation</td>\n",
       "      <td>not applicable</td>\n",
       "      <td>not applicable</td>\n",
       "      <td>D skin_armpit_left</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>73 rows × 76 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "         sample_name        bacteria bacteria_function biomass buffer_removal  \\\n",
       "1    14332.361158597  not applicable    not applicable     low            dry   \n",
       "3    14332.361162759  not applicable    not applicable     low            dry   \n",
       "4    14332.361162761  not applicable    not applicable     low            dry   \n",
       "11   14332.361162773  not applicable    not applicable     low            dry   \n",
       "21   14332.361162788  not applicable    not applicable     low            dry   \n",
       "..               ...             ...               ...     ...            ...   \n",
       "235  14332.363197876  not applicable    not applicable     low            dry   \n",
       "236  14332.363197877  not applicable    not applicable     low            dry   \n",
       "240  14332.363197882  not applicable    not applicable     low            dry   \n",
       "243  14332.363197890  not applicable    not applicable     low            dry   \n",
       "245  14332.363197892  not applicable    not applicable     low            dry   \n",
       "\n",
       "     col collection_timestamp compressed_well        description  \\\n",
       "1    8.0              3/24/22              O8    skin_hand_right   \n",
       "3    3.0              3/24/22              E3  skin_armpit_right   \n",
       "4    5.0              3/24/22              G5  skin_armpit_right   \n",
       "11   4.0              3/24/22              C4    skin_hand_right   \n",
       "21   6.0              3/24/22              G6    skin_hand_right   \n",
       "..   ...                  ...             ...                ...   \n",
       "235  8.0              3/24/22              H8     skin_hand_left   \n",
       "236  2.0              3/23/22              J2     skin_hand_left   \n",
       "240  7.0              3/24/22              D7   skin_armpit_left   \n",
       "243  8.0              3/24/22              D8     skin_hand_left   \n",
       "245  7.0              3/24/22              F7   skin_armpit_left   \n",
       "\n",
       "               diet  ...       sample_type2        scientific_name     sex  \\\n",
       "1    not applicable  ...    skin_hand_right  human skin metagenome  female   \n",
       "3    not applicable  ...  skin_armpit_right  human skin metagenome    male   \n",
       "4    not applicable  ...  skin_armpit_right  human skin metagenome  female   \n",
       "11   not applicable  ...    skin_hand_right  human skin metagenome    male   \n",
       "21   not applicable  ...    skin_hand_right  human skin metagenome  female   \n",
       "..              ...  ...                ...                    ...     ...   \n",
       "235  not applicable  ...     skin_hand_left  human skin metagenome  female   \n",
       "236  not applicable  ...     skin_hand_left  human skin metagenome  female   \n",
       "240  not applicable  ...   skin_armpit_left  human skin metagenome  female   \n",
       "243  not applicable  ...     skin_hand_left  human skin metagenome  female   \n",
       "245  not applicable  ...   skin_armpit_left  human skin metagenome  female   \n",
       "\n",
       "    storage_solution swab_type  taxon_id                       title  \\\n",
       "1        isopropanol     swube  539655.0  Matrix_pipeline_validation   \n",
       "3               etoh     swube  539655.0  Matrix_pipeline_validation   \n",
       "4               etoh     swube  539655.0  Matrix_pipeline_validation   \n",
       "11              etoh     swube  539655.0  Matrix_pipeline_validation   \n",
       "21              etoh     swube  539655.0  Matrix_pipeline_validation   \n",
       "..               ...       ...       ...                         ...   \n",
       "235      isopropanol    custom  539655.0  Matrix_pipeline_validation   \n",
       "236      isopropanol    custom  539655.0  Matrix_pipeline_validation   \n",
       "240             etoh    custom  539655.0  Matrix_pipeline_validation   \n",
       "243             etoh    custom  539655.0  Matrix_pipeline_validation   \n",
       "245             etoh    custom  539655.0  Matrix_pipeline_validation   \n",
       "\n",
       "           vivarium           water         replicate_id  \n",
       "1    not applicable  not applicable    D skin_hand_right  \n",
       "3    not applicable  not applicable  B skin_armpit_right  \n",
       "4    not applicable  not applicable  C skin_armpit_right  \n",
       "11   not applicable  not applicable    B skin_hand_right  \n",
       "21   not applicable  not applicable    C skin_hand_right  \n",
       "..              ...             ...                  ...  \n",
       "235  not applicable  not applicable     D skin_hand_left  \n",
       "236  not applicable  not applicable     A skin_hand_left  \n",
       "240  not applicable  not applicable   D skin_armpit_left  \n",
       "243  not applicable  not applicable     D skin_hand_left  \n",
       "245  not applicable  not applicable   D skin_armpit_left  \n",
       "\n",
       "[73 rows x 76 columns]"
      ]
     },
     "execution_count": 27,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Read in sample metadata\n",
    "md = pd.read_csv('{SEQ_TYPE}/metadata.tsv'.format(SEQ_TYPE=SEQ_TYPE), sep = '\\t')\n",
    "md = md[(md['description'] != 'BLANK') & (md['description'] != 'POSITIVE_CONTROL')]\n",
    "md = md[~(md['sample_name'].isin([\n",
    "    '14332.363197859', '14332.361164111',\n",
    "    '14332.363197857', '14332.363197849',\n",
    "    '14332.363197875', \n",
    "    '14332.363197831', '14332.363197893'\n",
    "\n",
    "]))]\n",
    "md = md[md.env_package == 'human-skin']\n",
    "md = md[md['env_feature'] == 'human-associated habitat']\n",
    "md.to_csv('shotgun/remove_blanks_md.tsv', sep='\\t', index=False)\n",
    "md"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Stepwise RDA"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Shotgun data"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 28,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Import data\n",
    "md_round1and2_bothPS_q2 = q2.Metadata.load('shotgun/remove_blanks_md.tsv')\n",
    "\n",
    "table_shotgun = q2.Artifact.load('shotgun/tree-filtered-table.qza')\n",
    "\n",
    "tree_shotgun = q2.Artifact.load('shotgun/tree.qza')\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 29,
   "metadata": {
    "scrolled": true
   },
   "outputs": [],
   "source": [
    "# Filter table\n",
    "table_shotgun_biom = table_shotgun.view(Table)\n",
    "md_round1and2_bothPS_df = md_round1and2_bothPS_q2.to_dataframe()\n",
    "shared_ = list(set(table_shotgun_biom.ids()) & set(md_round1and2_bothPS_df.index))\n",
    "md_round1and2_bothPS_df_shotgun = md_round1and2_bothPS_df.reindex(shared_)\n",
    "table_shotgun_biom_bothPS = table_shotgun_biom.filter(shared_)\n",
    "keep_ = table_shotgun_biom_bothPS.ids('observation')[table_shotgun_biom_bothPS.sum('observation') > 0]\n",
    "table_shotgun_biom_bothPS.filter(keep_, axis='observation')\n",
    "\n",
    "# Import filtered table and re-indexed metadata file\n",
    "table_shotgun_bothPS = q2.Artifact.import_data('FeatureTable[Frequency]', table_shotgun_biom_bothPS)\n",
    "md_round1and2_bothPS_df_shotgun_q2 = q2.Metadata(md_round1and2_bothPS_df_shotgun)\n",
    "\n",
    "# Generate distance matrices using 'all_dissts' utils\n",
    "rare_depth_shotgun = 1000\n",
    "dists_res_shotgun = all_dists(table_shotgun_bothPS,\n",
    "                      rare_depth_shotgun, tree_shotgun)\n",
    "\n",
    "# Generate ordinations (row=samples, cols=axes)\n",
    "pcoa_res_shotgun = {}\n",
    "pcoa_res_shotgun['Jaccard'] = pcoa(dists_res_shotgun['Jaccard'].distance_matrix).pcoa.view(OrdinationResults).samples\n",
    "pcoa_res_shotgun['Unweighted UniFrac'] = pcoa(dists_res_shotgun['Unweighted UniFrac'].distance_matrix).pcoa.view(OrdinationResults).samples\n",
    "pcoa_res_shotgun['Weighted UniFrac'] = pcoa(dists_res_shotgun['Weighted UniFrac'].distance_matrix).pcoa.view(OrdinationResults).samples\n",
    "pcoa_res_shotgun['RPCA'] = dists_res_shotgun['RPCA'].biplot.view(OrdinationResults).samples\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 30,
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "23\n",
      "Running external command line application(s). This may print messages to stdout and/or stderr.\n",
      "The command(s) being run are below. These commands cannot be manually re-run as they will depend on temporary files that no longer exist.\n",
      "\n",
      "Command: /home/kalen/matrix-tube-analysis/matrix-2.0/present/assets/stepwise-rda.R /tmp/tmpcxqvxa8v/ord_.tsv /tmp/tmpcxqvxa8v/mf_.txt /tmp/tmpcxqvxa8v/output.effect.size.tsv\n",
      "\n",
      "Running external command line application(s). This may print messages to stdout and/or stderr.\n",
      "The command(s) being run are below. These commands cannot be manually re-run as they will depend on temporary files that no longer exist.\n",
      "\n",
      "Command: /home/kalen/matrix-tube-analysis/matrix-2.0/present/assets/stepwise-rda.R /tmp/tmp4dk63md9/ord_.tsv /tmp/tmp4dk63md9/mf_.txt /tmp/tmp4dk63md9/output.effect.size.tsv\n",
      "\n",
      "Running external command line application(s). This may print messages to stdout and/or stderr.\n",
      "The command(s) being run are below. These commands cannot be manually re-run as they will depend on temporary files that no longer exist.\n",
      "\n",
      "Command: /home/kalen/matrix-tube-analysis/matrix-2.0/present/assets/stepwise-rda.R /tmp/tmp2dct9g09/ord_.tsv /tmp/tmp2dct9g09/mf_.txt /tmp/tmp2dct9g09/output.effect.size.tsv\n",
      "\n",
      "Running external command line application(s). This may print messages to stdout and/or stderr.\n",
      "The command(s) being run are below. These commands cannot be manually re-run as they will depend on temporary files that no longer exist.\n",
      "\n",
      "Command: /home/kalen/matrix-tube-analysis/matrix-2.0/present/assets/stepwise-rda.R /tmp/tmppii03kwb/ord_.tsv /tmp/tmppii03kwb/mf_.txt /tmp/tmppii03kwb/output.effect.size.tsv\n",
      "\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th>R2.adj</th>\n",
       "      <th>Df</th>\n",
       "      <th>AIC</th>\n",
       "      <th>F</th>\n",
       "      <th>Pr(&gt;F)</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th rowspan=\"2\" valign=\"top\">Jaccard</th>\n",
       "      <th>+ host_subject_id</th>\n",
       "      <td>0.197749</td>\n",
       "      <td>3</td>\n",
       "      <td>68.000576</td>\n",
       "      <td>6.915822</td>\n",
       "      <td>0.000200</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>+ extraction_protocol</th>\n",
       "      <td>0.027091</td>\n",
       "      <td>1</td>\n",
       "      <td>66.427146</td>\n",
       "      <td>3.411491</td>\n",
       "      <td>0.020196</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th rowspan=\"2\" valign=\"top\">Unweighted UniFrac</th>\n",
       "      <th>+ host_subject_id</th>\n",
       "      <td>0.174812</td>\n",
       "      <td>3</td>\n",
       "      <td>70.058383</td>\n",
       "      <td>6.084295</td>\n",
       "      <td>0.000200</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>+ extraction_protocol</th>\n",
       "      <td>0.040887</td>\n",
       "      <td>1</td>\n",
       "      <td>67.282892</td>\n",
       "      <td>4.597131</td>\n",
       "      <td>0.004199</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Weighted UniFrac</th>\n",
       "      <th>+ host_subject_id</th>\n",
       "      <td>0.143698</td>\n",
       "      <td>3</td>\n",
       "      <td>72.760282</td>\n",
       "      <td>5.027496</td>\n",
       "      <td>0.000200</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>RPCA</th>\n",
       "      <th>+ host_subject_id</th>\n",
       "      <td>0.203455</td>\n",
       "      <td>3</td>\n",
       "      <td>67.479509</td>\n",
       "      <td>7.130121</td>\n",
       "      <td>0.000200</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                                            R2.adj  Df        AIC         F  \\\n",
       "Jaccard            + host_subject_id      0.197749   3  68.000576  6.915822   \n",
       "                   + extraction_protocol  0.027091   1  66.427146  3.411491   \n",
       "Unweighted UniFrac + host_subject_id      0.174812   3  70.058383  6.084295   \n",
       "                   + extraction_protocol  0.040887   1  67.282892  4.597131   \n",
       "Weighted UniFrac   + host_subject_id      0.143698   3  72.760282  5.027496   \n",
       "RPCA               + host_subject_id      0.203455   3  67.479509  7.130121   \n",
       "\n",
       "                                            Pr(>F)  \n",
       "Jaccard            + host_subject_id      0.000200  \n",
       "                   + extraction_protocol  0.020196  \n",
       "Unweighted UniFrac + host_subject_id      0.000200  \n",
       "                   + extraction_protocol  0.004199  \n",
       "Weighted UniFrac   + host_subject_id      0.000200  \n",
       "RPCA               + host_subject_id      0.000200  "
      ]
     },
     "execution_count": 30,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Perform stepwise RDA-ANOVA\n",
    "es_all = {}\n",
    "use_ = ['host_subject_id', 'storage_solution', 'extraction_protocol']\n",
    "\n",
    "# Clean up meta (only stuff to run)\n",
    "mf_ord = md_round1and2_bothPS_df_shotgun_q2.to_dataframe().copy()\n",
    "\n",
    "# Filter data\n",
    "keep_ = [v_ for v_ in mf_ord.columns\n",
    "         if len(set(mf_ord[v_])) > 1 and\n",
    "         len(set(mf_ord[v_])) < mf_ord.shape[0]//2]\n",
    "mf_ord = mf_ord[keep_]\n",
    "print(len(keep_))\n",
    "# Run stepwise ANOVA for all RDA ordinations\n",
    "for metric_, ord_ in  pcoa_res_shotgun.items():\n",
    "    # get first three axes\n",
    "    ord_ = ord_[[0,1,2]]\n",
    "    ord_.columns = ['PC1','PC2','PC3']\n",
    "    # subset/match\n",
    "    mf_ord_ = mf_ord.copy()\n",
    "    shared_ids = list(set(ord_.index)\\\n",
    "                      & set(mf_ord_.index))\n",
    "    mf_ord_ = mf_ord_.loc[shared_ids,:]\n",
    "    ord_ = ord_.loc[shared_ids,:]\n",
    "    es_all[metric_] = run_stepwise_anova(ord_, mf_ord_, use_) #mf_ord_.columns)\n",
    "\n",
    "# Concat output from all runs and export\n",
    "es_alldf = pd.concat(es_all).rename({'+ sample_type2':'+ sample_type'}, axis=0)\n",
    "# es_alldf.to_csv('/Users/Justin/Mycelium/UCSD/00_Knight_Lab/03_Extraction_test_12201/round_02/results/stepwise_anova/stepwise_anova_shotgun.txt', sep='\\t')\n",
    "es_alldf\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 31,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "(73, 3)"
      ]
     },
     "execution_count": 31,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "pcoa_res_shotgun['RPCA'].shape"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.6.10"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 4
}
